### PR TITLE
optimization for eliminating redundant memory operations

### PR DESCRIPTION
**Pull request by yyzdtccjdtc**
Optimization for eliminating redundant memory operations related to this issue.
https://github.com/visioncortex/visioncortex/issues/3#issue-1230569501

### CODE REVIEW

**Comment by yyzdtccjdtc** [comment]
@tyt2y3 Would you like to give some comments about this pull request?

**Comment by tyt2y3** [comment]
Thank you for the contribution!
Actually, I never expected PR for this project.
Sadly we don't even have a CI process to test build, nor a release process to also test our downstreams (vtracer etc). I assume those projects can build without changing anything? Otherwise we might also need to update the other projects too.
I looked briefly, and it appears to be no harm. But I need to have some more confidence before merging in.

File: src/color_clusters/builder.rs [hunk 1]
```diff
@@ -134,7 +134,7 @@ struct Area {
     pub count: usize,
 }
 
-struct BuilderImpl {
+pub struct BuilderImpl {
```
Comment (tyt2y3) [inline src/color_clusters/builder.rs:137]: if possible prefer to `pub(crate)` instead of `pub`

File: src/color_clusters/builder.rs [hunk 2]
```diff
@@ -143,11 +143,11 @@ struct BuilderImpl {
     diff: Diff,
     deepen: Deepen,
     hollow: Hollow,
-    width: u32,
-    height: u32,
+    pub width: u32,
```
Comment (tyt2y3) [inline src/color_clusters/builder.rs]: same for these.

**Comment by tyt2y3** [comment]
OH I also noticed some file mode changes `100644 → 100755` which definitely is unnecessary

**Comment by yyzdtccjdtc** [comment]
@tyt2y3 I‘ve changed `width`, `height` and `cluster_indices` fields from `pub` to `pub(crate)`. But for 
`BuilderImpl` structure, I can't change it to `pub(crate)` because this structure will be used in some `pub` functions, like `perimeter` and `neighbours `, and if I change it to `pub`, the compiler will report an error.

Since I am using vtracer for testing, I am sure that changing this library will not have any effect on the running of this program. If you still have concerns about some `pub` functions, I can also write some new functions while keeping the original one, and then only call these new functions inside the library, so that it will never affect the use of other programs.

By the way, what do you mean by file mode changes?

**Comment by tyt2y3** [comment]
> what do you mean by file mode changes

image

> write some new functions while keeping the original one, and then only call these new functions inside the library, so that it will never affect the use of other programs

yeah that would be appreciated

**Comment by yyzdtccjdtc** [comment]
I've added some new pub(crate) functions to the latest commit that is guaranteed to speed up the program while not affecting other pub functions that may be called by the other programs. Please let me know if you have any other concerns about this pull request.

**Comment by tyt2y3** [comment]
Great thanks! Will try that out

**Pull request merged**